components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,387 @@
+use std::{boxed::Box, error::Error, iter::Iterator, slice::Iter, str::FromStr, vec::Vec};
+
+/// Represents the maximum Unicode Code Point, inclusive
+const CODEPOINT_MAX: u32 = 0x10FFFF;
+/// Represents the minimum Unicode Code Point, inclusive
+const CODEPOINT_MIN: u32 = 0x000000;
+/// Represents the end code point of the Basic Multilingual Plane range, starting from code point 0 , inclusive
+const BMP_MAX: u32 = 0xFFFF;
+
+/// Given string representation of inversion list create set
+///
+/// See UnicodeSet::new for conditions
+fn parse_serial_string(serialize_str: &str) -> Result<Vec<u32>, Box<dyn Error>> {
+    let mut serialize = serialize_str.split(" ");
+    let capacity: usize = serialize
+        .next()
+        .ok_or(format!(
+            "Splitting by spaces did not yield a capacity: {:?}",
+            serialize
+        ))?
+        .parse()?;
+    if capacity % 2 != 0 {
+        return Err(format!("Capacity must be even. Parsed Capacity: {}", capacity).into());
+    }
+    let mut serialized_vec: Vec<u32> = Vec::with_capacity(capacity);
+    let mut prev: Option<u32> = None;
+    for str_elem in serialize {
+        let parsed: u32 = str_elem.parse()?;
+        if serialized_vec.len() + 1 > serialized_vec.capacity() {
+            return Err(format!(
+                "Serialization capacity is too small. Allocated Capacity: {}",
+                capacity
+            )
+            .into());
+        }
+        if Some(parsed) <= prev {
+            return Err(format!(
+                "Serialization must be sorted. {:?} followed by {:?}",
+                Some(parsed),
+                prev
+            )
+            .into());
+        }
+        serialized_vec.push(parsed);
+        prev = Some(parsed);
+    }
+    if serialized_vec.len() % 2 != 0 {
+        return Err(format!(
+            "Serialization must be even. Serialization: {:?} Length: {}",
+            serialized_vec,
+            serialized_vec.len()
+        )
+        .into());
+    }
+    serialized_vec.shrink_to_fit(); // necessary if the length < capacity
+    Ok(serialized_vec)
+}
```

Comment:
I'm not sure how perf-critical it is, so assuming it is not and you can afford two vectors, here's a simpler code:

```rust
fn parse_serial_string(input: &str) -> Result<Vec<Range<u32>>, Box<dyn Error>> {
    let values = input
      .split(" ")
      .map(|s| s.parse())
      .collect::<Result<Vec<u32>, _>>()?;
    let chunks = values.chunks_exact(2);
    debug_assert_eq!(chunks.remainder().len(), 0);
    
    let pairs = chunks.map(|r| r[0]..r[1]).collect();
    Ok(pairs)
}
```

If it is perf-critical we can optimize the iterator directly to move through pairs. You can also debug assert that it's sorted.

Reply:
See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=391be83a14467f8628de0f5dbd6648aa